### PR TITLE
[GSoC] robust function for affine transform estimation

### PR DESCRIPTION
**Pull request by hrnr**
implements `estimateAffine2D`. estimates affine transformation using robust RANSAC method.

* uses RANSAC framework in calib3d
* includes accuracy test
* uses SVD decomposition for solving 3 point equation

things to consider:
* is solving through SVD fast enough? should we make it symmetric and use EIGEN val decomposition?
* is checking for coplanar points robust enough? (it seems so)

I plan to write perf test to see if make sense to do some tricks and solve with EIGEN val decomp. Checking for coplanar points seems to work well for me. I adapted the check from estimateAffine3D, estimateRigidTransform have some addition checks (checking if points are not too close), but I'm not sure if it is worth to implement those.

I will implement also `estimateAffinePartial2D` that will estimate transformation using only rotation, translation an scaling (similar to `estimateRigidTransform`). I'm only not sure about the name `estimateAffinePartial2D`, any ideas welcome.

cc: @prclibo

### CODE REVIEW

**Comment by hrnr** [comment]
I don't have any idea why this fails OpenCL perf test. It does not make any sense to me. Is that some bug in `OCL_PyrLKOpticalFlowFixture_PyrLKOpticalFlow.PyrLKOpticalFlow`?

**Comment by hrnr** [comment]
added `estimateAffinePartial2D` and rebased.

File: modules/calib3d/include/opencv2/calib3d.hpp [hunk 1]
```diff
@@ -1548,6 +1548,53 @@ CV_EXPORTS_W  int estimateAffine3D(InputArray src, InputArray dst,
...
+@param dst Second input 2D point set.
+@param out Output 2D affine transformation (5 degrees of freedom) matrix \f$2 \times 3\f$ .
+@param inliers Output vector indicating which points are inliers.
+@param ransacThreshold Maximum reprojection error in the RANSAC algorithm to consider a point as
+an inlier.
+@param confidence Confidence level, between 0 and 1, for the estimated transformation. Anything
+between 0.95 and 0.99 is usually good enough. Values too close to 1 can slow down the estimation
+significantly. Values lower than 0.8-0.9 can result in an incorrectly estimated transformation.
+
+The function estimates an optimal 2D affine transformation with 5 degrees of freedom limited to
```
Comment (prclibo) [inline modules/calib3d/include/opencv2/calib3d.hpp]: do you mean 4 dof? 1 rotation + 1 scale + 2 translation?

**Comment by hrnr** [comment]
added perf test for both functions. Here are my results:

```
Time compensation is 0
CTEST_FULL_OUTPUT
OpenCV version: 3.1.0-dev
OpenCV VCS version: 3.1.0-663-g9e32c5c
Build type: release
Parallel framework: pthreads
CPU features: mmx sse sse2 sse3
OpenCL is disabled
Note: Google Test filter = *Estimate*
[==========] Running 18 tests from 2 test cases.
[----------] Global test environment set-up.
[----------] 9 tests from EstimateAffine_EstimateAffine2D
[ RUN      ] EstimateAffine_EstimateAffine2D.EstimateAffine2D/0
[ PERFSTAT ]    (samples = 10, mean = 2.93, median = 2.91, stddev = 0.07 (2.5%))
[ VALUE    ]    (5000, 0.99)
[       OK ] EstimateAffine_EstimateAffine2D.EstimateAffine2D/0 (30 ms)
[ RUN      ] EstimateAffine_EstimateAffine2D.EstimateAffine2D/1
[ PERFSTAT ]    (samples = 10, mean = 1.87, median = 1.87, stddev = 0.02 (0.9%))
[ VALUE    ]    (5000, 0.95)
[       OK ] EstimateAffine_EstimateAffine2D.EstimateAffine2D/1 (19 ms)
[ RUN      ] EstimateAffine_EstimateAffine2D.EstimateAffine2D/2
[ PERFSTAT ]    (samples = 10, mean = 1.46, median = 1.45, stddev = 0.02 (1.2%))
[ VALUE    ]    (5000, 0.9)
[       OK ] EstimateAffine_EstimateAffine2D.EstimateAffine2D/2 (14 ms)
[ RUN      ] EstimateAffine_EstimateAffine2D.EstimateAffine2D/3
[ PERFSTAT ]    (samples = 10, mean = 0.73, median = 0.72, stddev = 0.01 (1.5%))
[ VALUE    ]    (1000, 0.99)
[       OK ] EstimateAffine_EstimateAffine2D.EstimateAffine2D/3 (8 ms)
[ RUN      ] EstimateAffine_EstimateAffine2D.EstimateAffine2D/4
[ PERFSTAT ]    (samples = 10, mean = 0.48, median = 0.47, stddev = 0.01 (2.9%))
[ VALUE    ]    (1000, 0.95)
[       OK ] EstimateAffine_EstimateAffine2D.EstimateAffine2D/4 (5 ms)
[ RUN      ] EstimateAffine_EstimateAffine2D.EstimateAffine2D/5
[ PERFSTAT ]    (samples = 10, mean = 0.41, median = 0.41, stddev = 0.01 (2.4%))
[ VALUE    ]    (1000, 0.9)
[       OK ] EstimateAffine_EstimateAffine2D.EstimateAffine2D/5 (4 ms)
[ RUN      ] EstimateAffine_EstimateAffine2D.EstimateAffine2D/6
[ PERFSTAT ]    (samples = 13, mean = 0.24, median = 0.24, stddev = 0.01 (2.1%))
[ VALUE    ]    (100, 0.99)
[       OK ] EstimateAffine_EstimateAffine2D.EstimateAffine2D/6 (3 ms)
[ RUN      ] EstimateAffine_EstimateAffine2D.EstimateAffine2D/7
[ PERFSTAT ]    (samples = 17, mean = 0.16, median = 0.15, stddev = 0.00 (3.0%))
[ VALUE    ]    (100, 0.95)
[       OK ] EstimateAffine_EstimateAffine2D.EstimateAffine2D/7 (3 ms)
[ RUN      ] EstimateAffine_EstimateAffine2D.EstimateAffine2D/8
[ PERFSTAT ]    (samples = 13, mean = 0.13, median = 0.13, stddev = 0.00 (1.2%))
[ VALUE    ]    (100, 0.9)
[       OK ] EstimateAffine_EstimateAffine2D.EstimateAffine2D/8 (2 ms)
[----------] 9 tests from EstimateAffine_EstimateAffine2D (88 ms total)

[----------] 9 tests from EstimateAffine_EstimateAffinePartial2D
[ RUN      ] EstimateAffine_EstimateAffinePartial2D.EstimateAffinePartial2D/0
[ PERFSTAT ]    (samples = 10, mean = 1.07, median = 1.05, stddev = 0.03 (2.9%))
[ VALUE    ]    (5000, 0.99)
[       OK ] EstimateAffine_EstimateAffinePartial2D.EstimateAffinePartial2D/0 (10 ms)
[ RUN      ] EstimateAffine_EstimateAffinePartial2D.EstimateAffinePartial2D/1
[ PERFSTAT ]    (samples = 10, mean = 0.70, median = 0.69, stddev = 0.01 (1.6%))
[ VALUE    ]    (5000, 0.95)
[       OK ] EstimateAffine_EstimateAffinePartial2D.EstimateAffinePartial2D/1 (8 ms)
[ RUN      ] EstimateAffine_EstimateAffinePartial2D.EstimateAffinePartial2D/2
[ PERFSTAT ]    (samples = 10, mean = 0.54, median = 0.53, stddev = 0.01 (2.0%))
[ VALUE    ]    (5000, 0.9)
[       OK ] EstimateAffine_EstimateAffinePartial2D.EstimateAffinePartial2D/2 (5 ms)
[ RUN      ] EstimateAffine_EstimateAffinePartial2D.EstimateAffinePartial2D/3
[ PERFSTAT ]    (samples = 25, mean = 0.25, median = 0.25, stddev = 0.01 (2.7%))
[ VALUE    ]    (1000, 0.99)
[       OK ] EstimateAffine_EstimateAffinePartial2D.EstimateAffinePartial2D/3 (7 ms)
[ RUN      ] EstimateAffine_EstimateAffinePartial2D.EstimateAffinePartial2D/4
[ PERFSTAT ]    (samples = 13, mean = 0.17, median = 0.16, stddev = 0.00 (2.8%))
[ VALUE    ]    (1000, 0.95)
[       OK ] EstimateAffine_EstimateAffinePartial2D.EstimateAffinePartial2D/4 (2 ms)
[ RUN      ] EstimateAffine_EstimateAffinePartial2D.EstimateAffinePartial2D/5
[ PERFSTAT ]    (samples = 13, mean = 0.13, median = 0.13, stddev = 0.00 (2.9%))
[ VALUE    ]    (1000, 0.9)
[       OK ] EstimateAffine_EstimateAffinePartial2D.EstimateAffinePartial2D/5 (2 ms)
[ RUN      ] EstimateAffine_EstimateAffinePartial2D.EstimateAffinePartial2D/6
[ PERFSTAT ]    (samples = 36, mean = 0.07, median = 0.07, stddev = 0.00 (3.0%))
[ VALUE    ]    (100, 0.99)
[       OK ] EstimateAffine_EstimateAffinePartial2D.EstimateAffinePartial2D/6 (2 ms)
[ RUN      ] EstimateAffine_EstimateAffinePartial2D.EstimateAffinePartial2D/7
[ PERFSTAT ]    (samples = 25, mean = 0.05, median = 0.05, stddev = 0.00 (2.9%))
[ VALUE    ]    (100, 0.95)
[       OK ] EstimateAffine_EstimateAffinePartial2D.EstimateAffinePartial2D/7 (1 ms)
[ RUN      ] EstimateAffine_EstimateAffinePartial2D.EstimateAffinePartial2D/8
[ PERFSTAT ]    (samples = 25, mean = 0.04, median = 0.04, stddev = 0.00 (2.8%))
[ VALUE    ]    (100, 0.9)
[       OK ] EstimateAffine_EstimateAffinePartial2D.EstimateAffinePartial2D/8 (1 ms)
[----------] 9 tests from EstimateAffine_EstimateAffinePartial2D (38 ms total)

[----------] Global test environment tear-down
[==========] 18 tests from 2 test cases ran. (126 ms total)
[  PASSED  ] 18 tests.
```

`estimateAffinePartial2D` is about 3 times faster. This is expected result since it needs only 2 points for model (instead of 3) and the system to solve is also smaller.

File: modules/calib3d/include/opencv2/calib3d.hpp (hunk 1 above)
Comment (hrnr) [inline modules/calib3d/include/opencv2/calib3d.hpp]: yes. I will fix it.

**Comment by hrnr** [comment]
Is it normal that perf test fails on a first run? It fails with `No regression data for aff_est argument`, that must be because of `SANITY_CHECK` macro.

It does not happen locally.

**Comment by hrnr** [comment]
added regression data in Itseez/opencv_extra#296. Hope that it will make build bot happy.

**Comment by hrnr** [comment]
I have experimented with using eigenval decomposition instead of SVD in eba30a89737d4ded755f07cff75fd861864cf09a (+ trick to make the matrix symmetric), but it is not any faster.

From my POV I consider this feature complete. If we convice the buildbot to use the new generated regression data I can rebase and this could be the first thing to go in.

Since this blocks #6609, I will focus on parallel feature finding subtask for now and continue with affine registration in sttiching pipeline once this will go in.

**Comment by prclibo** [comment]
Basically looks ok to me. 

the test size of (100000, 5000, 100) is different from what you have in the sanity sheet in  Itseez/opencv_extra#296. Do you need to regenerate it?

**Comment by hrnr** [comment]
Thanks for noticing, I have regenerated the data. Hope that buildbot will be happy now.

I have rebased my commits.

**Comment by hrnr** [comment]
Buildbot is failing to merge opencv_extra with my regression data. Is there anything I can do to convice buildbot to use the generated data?

**Comment by alalek** [comment]
You should use the same name for branch in opencv_extra: `estimate_affine_2d`

**Comment by hrnr** [comment]
Thanks, reopened as Itseez/opencv_extra#297 with correct branch name.

Rebased against newest master to force builtbot run.

**Comment by hrnr** [comment]
All green. For me this is good to go in.

**Comment by hrnr** [comment]
Is everything OK with this? Am I supposed to take some action to get this merged?

**Comment by prclibo** [comment]
@alalek , this function is an independent module from Jiri's project and I think it looks completed. Can you take a look and merge this pull first so it would be easier for Jiri to send pulls of another modules based on this?

**Comment by hrnr** [comment]
incorporated fix from #6768

**Comment by vpisarev** [comment]
this PR will be reopened because of CI issues

**Comment by phryniszak** [comment]
estimateAffinePartial2D and estimateAffine2D seems to be buggy. Fails for int input values. Returned values are cast to int so results makes not sens.
More details here:
http://answers.opencv.org/question/123476/estimateaffinepartial2d-and-estimateaffine2d/

**Comment by hrnr** [comment]
In case estimateAffine* functions get something different than floats, inputs are converted to floats and then algorithm works the same. Returned matrix is always doubles.

estimateAffine* uses Levenberg-Marquart as final refinement step. Do you get expected results when you call it with `refineIters=0`?

**Comment by phryniszak** [comment]
I tried with `refineIters=0` and results are still the same.

**Comment by hrnr** [comment]
Thanks for report, I will take a look on that. There might be some degenerate points configuration not catched by tests.

I see you are now storing `cv::Point2f` into `cv::Point`. Are you sure conversion does what you expect?

**Comment by phryniszak** [comment]
std::vector<cv::Point> p1s,p2s and results are:
```
estimateRigidTransform false: [0.7100000000000001, 0.7100000000000001, 0;
 -0.7100000000000001, 0.7100000000000001, 0]
estimateRigidTransform true:[0.7100000000000001, 0.7100000000000001, 0;
 -0.7100000000000001, 0.7100000000000001, 0]
estimateAffine2D[1, -0, -0;
 -0, 1, -0]
estimateAffinePartial2D[1, -0, 0;
 0, 1, 0]

```
and std::vector<cv::Point2f> p1s,p2s; and correct results: 
```

stimateRigidTransform false: [0.7100000000000001, 0.7100000000000001, 0;
 -0.7100000000000001, 0.7100000000000001, 0]
estimateRigidTransform true:[0.7100000000000001, 0.7100000000000001, 0;
 -0.7100000000000001, 0.7100000000000001, 0]
estimateAffine2D[0.7100000000000001, 0.7100000000000001, -0;
 -0.7100000000000001, 0.7100000000000001, -0]
estimateAffinePartial2D[0.7100000000000001, 0.7100000000000001, 0;
 -0.7100000000000001, 0.7100000000000001, 0]
```

**Comment by hrnr** [comment]
You are right, there is a stupid bug in estimateAffine* affecting conversion of inputs.

I will fix that (please give me a few days, I'm quite busy now). Thank you for reporting this, we should have probably more tests to cover also conversions.

**Pull request closed without merging**